File: apps/bank/verify-transaction/java/src/main/java/com/app/VerifyTransaction.java
package com.app;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Future;
import io.vertx.core.Promise;
import io.vertx.core.Vertx;
import io.vertx.core.MultiMap;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.core.buffer.Buffer;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.RoutingContext;
import io.vertx.ext.web.handler.BodyHandler;
import io.vertx.ext.web.client.HttpRequest;
import io.vertx.ext.web.client.WebClient;
import io.vertx.ext.web.client.WebClientOptions;

import java.lang.reflect.Array;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Collectors;

public class VerifyTransaction extends AbstractVerticle {

  // Required environment variables
  private static final String DIRECTUS_URL = System.getenv("DIRECTUS_URL");
  private static final String DIRECTUS_TOKEN = System.getenv("DIRECTUS_TOKEN");
  private static final String K_SINK = System.getenv("K_SINK");
  private static final String OTP_SECRET = System.getenv("OTP_SECRET");
  private static final double MAX_AMOUNT = System.getenv("MAX_AMOUNT") != null
      ? Double.parseDouble(System.getenv("MAX_AMOUNT"))
      : 0;

  // Headers to remove from incoming request when forwarding.
  private static final String[] HEADERS_REMOVE = new String[]{
      "ce-id", "ce-specversion", "ce-type", "ce-source", "content-type", "host", "x-k-sink"
  };

  private WebClient webClient;

  @Override
  public void start(Promise<Void> startPromise) {
    if (DIRECTUS_URL == null || DIRECTUS_TOKEN == null || OTP_SECRET == null || MAX_AMOUNT == 0) {
      startPromise.fail("Missing required environment variables: DIRECTUS_URL, DIRECTUS_TOKEN, MAX_AMOUNT, OTP_SECRET");
      return;
    }
    if (K_SINK == null) {
      System.out.println("Warning: K_SINK is not set. Events will not be forwarded.");
    }

    webClient = WebClient.create(vertx, new WebClientOptions().setSsl(false));

    Router router = Router.router(vertx);
    router.route().handler(BodyHandler.create());
    router.post("/").handler(this::handleTransaction);

    vertx.createHttpServer()
         .requestHandler(router)
         .listen(8080, http -> {
           if (http.succeeded()) {
             startPromise.complete();
             System.out.println("Server started on port 8080");
           } else {
             startPromise.fail(http.cause());
           }
         });
  }

  private void handleTransaction(RoutingContext ctx) {
    JsonObject requestBody = ctx.getBodyAsJson();
    if (requestBody == null) {
      forwardToBroker(ctx, new JsonObject(), false, "Invalid JSON", 400);
      return;
    }
    
    // Use atomic containers to allow modifications within lambdas.
    final AtomicBoolean proceedRef = new AtomicBoolean(true);
    final AtomicReference<String> messageRef = new AtomicReference<>("Forwarded");
    final AtomicInteger statusCodeRef = new AtomicInteger(200);
    double amount = 0;

    // Parse "amount"
    try {
      Object amtObj = requestBody.getValue("amount");
      if (amtObj instanceof Number) {
        amount = ((Number) amtObj).doubleValue();
      } else if (amtObj instanceof String) {
        amount = Double.parseDouble((String) amtObj);
      } else {
        throw new Exception("Invalid type");
      }
    } catch (Exception e) {
      proceedRef.set(false);
      messageRef.set("Invalid amount provided");
      statusCodeRef.set(400);
    }

    String clientUsername = requestBody.getString("client");
    String destinationClient = requestBody.getString("destination_client");
    String otp = requestBody.getString("otp");

    // Check OTP if amount exceeds MAX_AMOUNT.
    if (statusCodeRef.get() == 200 && amount > MAX_AMOUNT && (otp == null || !otp.equals(OTP_SECRET))) {
      proceedRef.set(false);
      messageRef.set("OTP required or incorrect");
      statusCodeRef.set(403);
    }

    String req_id = ctx.request().getHeader("Ce-Id");

    // Synchronously retrieve user IDs for both client and destination_client.
    Future<Integer> fromFuture = getUserId(clientUsername, req_id);
    Future<Integer> toFuture = fromFuture.compose(clientId -> getUserId(destinationClient, req_id));

    io.vertx.core.CompositeFuture.all(fromFuture, toFuture).onComplete(ar -> {
      Integer fromId = fromFuture.result();
      Integer toId = toFuture.result();
      if (statusCodeRef.get() == 200 && (fromId == null || toId == null)) {
        proceedRef.set(false);
        messageRef.set("Invalid client usernames");
        statusCodeRef.set(400);
      }
      // Build payload with additional fields.
      JsonObject payload = requestBody.copy();
      payload.put("message", messageRef.get());
      payload.put("from", fromId);
      payload.put("to", toId);
      forwardToBroker(ctx, payload, proceedRef.get(), messageRef.get(), statusCodeRef.get());
    });
  }

  /**
   * Retrieves a user ID from Directus given a username.
   */
  private Future<Integer> getUserId(String username, String req_id) {
    Promise<Integer> promise = Promise.promise();
    if (username == null) {
      promise.complete(null);
      return promise.future();
    }
    String url = DIRECTUS_URL + "/items/users?filter[username][_eq]=" + username;
    webClient.getAbs(url)
             .putHeader("Authorization", "Bearer " + DIRECTUS_TOKEN)
             .putHeader("Ce-Id", req_id)
             .send(ar -> {
               if (ar.succeeded() && ar.result().statusCode() == 200) {
                 JsonObject body = ar.result().bodyAsJsonObject();
                 if (body != null) {
                   JsonArray data = body.getJsonArray("data");
                   if (data != null && !data.isEmpty()) {
                     Integer id = data.getJsonObject(0).getInteger("id");
                     promise.complete(id);
                     return;
                   }
                 }
               }
               promise.complete(null);
             });
    return promise.future();
  }

  /**
   * Forwards the transaction payload to the external broker (K_SINK) and responds to the original requester.
   * Outgoing headers are constructed to prevent overrides by incoming headers.
   */
  private void forwardToBroker(RoutingContext ctx, JsonObject payload, boolean proceed, String message, int statusCode) {
    MultiMap headers = MultiMap.caseInsensitiveMultiMap();

    // Copy allowed incoming headers
    for (Map.Entry<String, String> entry : ctx.request().headers()) {
        String key = entry.getKey();
        if (!Arrays.asList(HEADERS_REMOVE).contains(key.toLowerCase())) {
            headers.set(key, entry.getValue()); // Override any incorrect headers
        }
    }

    // Set additional headers
    headers.set("Ce-Id", ctx.request().getHeader("Ce-Id"));
    headers.set("Ce-Specversion", "1.0");
    headers.set("Ce-Type", "transaction");
    headers.set("Ce-Source", "verify-transaction");
    headers.set("Ce-Dt", String.valueOf(proceed).toLowerCase());
    headers.set("Content-Type", "application/json"); // Ensure this is set

    // Log headers to confirm they are being set
    headers.forEach(entry -> System.out.println("Setting Header: " + entry.getKey() + " = " + entry.getValue()));

    // Determine target sink
    String targetSink = ctx.request().getHeader("X-K-Sink");
    if (targetSink == null || targetSink.isEmpty()) {
        targetSink = K_SINK;
    }

    if (targetSink != null) {
        HttpRequest<Buffer> request = webClient.postAbs(targetSink);

        // Set headers explicitly
        headers.forEach(entry -> request.putHeader(entry.getKey(), entry.getValue()));

        request.sendJsonObject(payload).onComplete(ar -> {
            if (ar.succeeded()) {
                System.out.println("Successfully forwarded event");
            } else {
                System.out.println("Failed to forward event" + " Error: " + ar.cause().getMessage());
            }

            // Prevent duplicate response
            if (!ctx.response().ended()) {
                JsonObject responsePayload = new JsonObject().put("message", message);
                ctx.response()
                   .setStatusCode(statusCode)
                   .putHeader("Content-Type", "application/json")
                   .end(responsePayload.encode());
            }
        });
      } else {
          System.out.println("Warning: No valid sink configured. Event not forwarded.");
          if (!ctx.response().ended()) {
              JsonObject responsePayload = new JsonObject().put("message", message);
              ctx.response()
                .setStatusCode(statusCode)
                .putHeader("Content-Type", "application/json")
                .end(responsePayload.encode());
          }
      }
  }



  public static void main(String[] args) {
    Vertx vertx = Vertx.vertx();
    vertx.deployVerticle(new VerifyTransaction());
  }
}

File: apps/bank/verify-transaction/php/verify_transaction.php
<?php

use Swoole\Http\Server;
use Swoole\Http\Request;
use Swoole\Http\Response;
use GuzzleHttp\Client;

require 'vendor/autoload.php';

// Required environment variables
$DIRECTUS_URL = getenv('DIRECTUS_URL');
$DIRECTUS_TOKEN = getenv('DIRECTUS_TOKEN');
$K_SINK = getenv('K_SINK');
$OTP_SECRET = getenv('OTP_SECRET');
$MAX_AMOUNT = getenv('MAX_AMOUNT') ? (float)getenv('MAX_AMOUNT') : 0;

$HEADERS_REMOVE = [
    'ce-id', 'ce-specversion', 'ce-type', 'ce-source', 'content-type', 'host', 'x-k-sink'
];

if (!$DIRECTUS_URL || !$DIRECTUS_TOKEN || !$OTP_SECRET || $MAX_AMOUNT == 0) {
    die("Missing required environment variables: DIRECTUS_URL, DIRECTUS_TOKEN, MAX_AMOUNT, OTP_SECRET\n");
}
if (!$K_SINK) {
    echo "Warning: K_SINK is not set. Events will not be forwarded.\n";
}

$server = new Server("0.0.0.0", 8080);
$client = new Client();

$server->on("request", function (Request $request, Response $response) use ($client, $DIRECTUS_URL, $DIRECTUS_TOKEN, $K_SINK, $OTP_SECRET, $MAX_AMOUNT, $HEADERS_REMOVE) {
    // Only accept POST requests
    if ($request->server['request_method'] !== 'POST') {
        $response->status(405);
        $response->end(json_encode(["message" => "Method Not Allowed"]));
        return;
    }
    
    $body = json_decode($request->rawContent(), true);

    // Initialize variables to hold the processing state
    $proceed = true;
    $message = "Forwarded";
    $statusCode = 200;

    // If the JSON payload is invalid, update our state
    if (!$body) {
        $proceed = false;
        $message = "Invalid JSON";
        $statusCode = 400;
        $payload = [];
        forwardToBroker($request, $response, $payload, $proceed, $message, $statusCode, $K_SINK, $HEADERS_REMOVE, $client);
        return;
    }

    // Parse and validate "amount"
    $amount = 0;
    if (isset($body['amount'])) {
        if (is_numeric($body['amount'])) {
            $amount = (float)$body['amount'];
        } else {
            // Try to trim and parse; if still not valid, mark as error
            $amount = (float)trim($body['amount']);
            if (!is_numeric($amount)) {
                $proceed = false;
                $message = "Invalid amount provided";
                $statusCode = 400;
            }
        }
    } else {
        $proceed = false;
        $message = "Invalid amount provided";
        $statusCode = 400;
    }

    // Retrieve client usernames and OTP from the payload
    $clientUsername = $body['client'] ?? null;
    $destinationClient = $body['destination_client'] ?? null;
    // Cast the OTP value from the body to string for proper comparison
    $otp = isset($body['otp']) ? (string)$body['otp'] : null;
    // Cast the OTP_SECRET to string as well (though getenv() returns a string by default)
    $OTP_SECRET = (string)$OTP_SECRET;

    // Check OTP if the amount exceeds MAX_AMOUNT
    if ($statusCode === 200 && $amount > $MAX_AMOUNT && (empty($otp) || $otp !== $OTP_SECRET)) {
        $proceed = false;
        $message = "OTP required or incorrect";
        $statusCode = 403;
    }

    // Retrieve the request ID from the headers
    $req_id = $request->header['ce-id'] ?? '';

    // Synchronously retrieve user IDs for both the client and the destination client
    $fromId = getUserId($clientUsername, $req_id, $DIRECTUS_URL, $DIRECTUS_TOKEN, $client);
    $toId = getUserId($destinationClient, $req_id, $DIRECTUS_URL, $DIRECTUS_TOKEN, $client);

    if ($statusCode === 200 && ($fromId === null || $toId === null)) {
        $proceed = false;
        $message = "Invalid client usernames";
        $statusCode = 400;
    }

    // Build the payload by copying the original body and adding extra fields
    $payload = $body;
    $payload['message'] = $message;
    $payload['from'] = $fromId;
    $payload['to'] = $toId;

    // Forward to the broker and send the final response
    forwardToBroker($request, $response, $payload, $proceed, $message, $statusCode, $K_SINK, $HEADERS_REMOVE, $client);
});

function getUserId($username, $req_id, $DIRECTUS_URL, $DIRECTUS_TOKEN, $client) {
    if (!$username) return null;
    
    $url = "$DIRECTUS_URL/items/users?filter[username][_eq]=$username";
    try {
        $res = $client->request('GET', $url, [
            'headers' => [
                'Authorization' => "Bearer $DIRECTUS_TOKEN",
                'Ce-Id' => $req_id
            ]
        ]);
        $data = json_decode($res->getBody()->getContents(), true);
        return $data['data'][0]['id'] ?? null;
    } catch (Exception $e) {
        return null;
    }
}

function forwardToBroker($request, $response, $payload, $proceed, $message, $statusCode, $K_SINK, $HEADERS_REMOVE, $client) {
    $headers = [];
    // Copy allowed incoming headers
    foreach ($request->header as $key => $value) {
        if (!in_array(strtolower($key), $HEADERS_REMOVE)) {
            $headers[$key] = $value;
        }
    }
    
    // Set additional headers explicitly
    $headers['Ce-Id'] = $request->header['ce-id'] ?? '';
    $headers['Ce-Specversion'] = "1.0";
    $headers['Ce-Type'] = "transaction";
    $headers['Ce-Source'] = "verify-transaction";
    $headers['Ce-Dt'] = $proceed ? 'true' : 'false';
    $headers['Content-Type'] = "application/json";

    // Determine target sink from header or environment variable
    $targetSink = $request->header['x-k-sink'] ?? $K_SINK;

    if ($targetSink) {
        try {
            $client->request('POST', $targetSink, [
                'headers' => $headers,
                'json' => $payload
            ]);
            echo "Successfully forwarded event\n";
        } catch (Exception $e) {
            echo "Failed to forward event: " . $e->getMessage() . "\n";
        }
    } else {
        echo "Warning: No valid sink configured. Event not forwarded.\n";
    }
    
    // Respond to the original requester
    $response->status($statusCode);
    $response->header("Content-Type", "application/json");
    $response->end(json_encode(["message" => $message]));
}

$server->start();
?>
